src/pow.rs

```diff
@@ -381,218 +322,227 @@ fn count_leading_zeros(data: &[u8]) -> u8 {
     leading_zeros
 }
 
-/// The faucet will dynamically adjust the difficulty based on the
-/// current balance of the faucet to make it increasingly difficult
-/// to retrieve funds from the faucet. The actual equation for this
-/// is:
-///
-/// ```math
-/// f_unclamped(x)=(M-m)(1-\log_{q}\frac{x}{b})+m
-/// f_clamped(x) = \max(m,\min(M,f_unclamped(x)))
-/// f(x) = f_clamped(x)
-/// ```
-///
-/// where:
-///
-/// - `x` is the current balance in BTC
-/// - `M` is the maximum difficulty
-/// - `m` is the minimum difficulty
-/// - `b` is the minimum balance in BTC
-/// - `q` is the amount emitted per request in BTC
-///
-/// # Guarantees
-///
-/// This function guarantees that the difficulty will be between `min_difficulty`
-/// and `max_difficulty` given that the correctness assumptions are met.
-///
-/// # Correctness
-///
-/// For this function correctly output, you must ensure:
-///
-/// - `per_emission` > `Amount::ONE_BTC`, ideally >2 BTC due to the way the curve
-///   functions
-/// - `min_difficulty` <= `max_difficulty`
-/// - `max_difficulty`, `min_difficulty`, `min_balance`, `q` and `x` all > 0
-///
-/// # Expected clamping points
-///
-/// f(x) = M at x = b
-/// f(x) = m at x = bq.
-pub fn calculate_difficulty(x: f32, big_m: f32, m: f32, b: f32, q: f32) -> f32 {
-    // optimisation for when the balance is less than or equal to the min balance
-    if x <= b {
-        return big_m;
+pub struct DifficultyConfig {
+    big_m: u8,
+    m: u8,
+    b: f32,
+    /// Optimisation for when x >= b+Lq, which should be the majority of the time
+    min_diff_start: f32,
+    /// Optimisation for the linear function. This is the gradient of the linear function.
+    precompute_big_a: f32,
+    /// Optimisation for the linear function. This is the y-intercept of the linear function.
+    precompute_big_b: f32,
+}
+
+impl DifficultyConfig {
+    pub fn new(
+        max_diff: u8,
+        min_diff: u8,
+        min_balance: Amount,
+        per_claim: Amount,
+        difficulty_increase_coeff: f32,
+    ) -> Result<Self, DifficultyConfigError> {
+        if max_diff < min_diff {
+            return Err(DifficultyConfigError::MaxDiffMustBeGreaterThanMinDiff);
+        }
+
+        let big_m = max_diff as f32;
+        let m = min_diff as f32;
+        let b = min_balance.to_sat() as f32;
+        let q = per_claim.to_sat() as f32;
+        let big_l = difficulty_increase_coeff;
+
+        let min_diff_start = b + big_l * q;
+
+        let precompute_big_a = (m - big_m) / (big_l * q);
```

Comment:
Either do `.checked_div` instead of `/` or make sure that `q` is NEVER zero. You choose.